source/develop/release-management/features/virt/vm-lifecycle-with-kubevirt.md

```diff
@@ -0,0 +1,130 @@
+4. Most of the VMs monitoring remains the same.  
+5. Most of the VM related operations are still done directly with VDSM (rather than through Kubervirt).
+6. If we fail to run a VM on a node several times in a short-period of time, the node should be non-operational (and thus not be scheduled until fixed).
+
+## Milestone 3: Dynamic Cluster State Managed by Kubevirt
+
+### Goal
+At this point we have basic monitoring of VMs. This milestone is about detecting more changes in running VMs in Kubevirt and report them to oVirt. Ideally, most of VmDynamic would be managed in Kubevirt and will be in oVirt only for display purpose. This allows us to switch VM management operations (shutdown, migrate, suspend, ..) to be done via Kubervirt.
+
+### Design
+1. Receive the VMs state from the node and updating the dynamic part in the VM representation.  
+2. Starting highly-available VMs in Kubevirt.
+3. Change VM management operations: hibernate, restore, resume, migrate, shutdown, reboot, cancel-migration, power-off.
+
+### Benefits
+1. Kubevirt becomes the owner of the runtime state of the cluster while oVirt manages its static state (host-deployments, VM templates, import VMs from external providers, and so on).  
+2. Many VM operations are delegated to Kubevirt.
+
+## Future Changes
+
+Future changes depend on having the knowledge of storage and network resources in Kubevirt and thus harder to predict.  
+
+## VM Representation
+
+I believe that the best way to represent the VM is in JSON as follow:  
+1. Scheduling data: the cluster-related data mentioned before, in the form of properties in JSON.  
+2. VM dynamic data: things like status of the VM, run_on_vds, console ip, and so on. Initially they should be set with default values. Once the monitoring in Kubevirt receives them from the node, it should update the VM representation and some component needs to listen to these changes and report them to the engine in oVirt. Note that we may want to report some of the dynamic data that exist today, like the applications list, in a different way.   
+3. VM configuration: the static configuration of the VM, including its devices.  
+
+In oVirt we are in the process of representing the VM configuration in XML (domxml) to eliminate the duplicate conversion of the data (once in the engine, converting the internal representation into a dictionary, and once in VDSM, converting the dictionary into domxml). We would prefer to keep this approach and minimize data conversions. In light of this approach, there are few options of how to represent the VM configuration:  
+1. To generate the VM representation in JSON. There is a proposal (on-going work?) to implement another binding for libvirt that will enable to pass the VM definition in JSON. With this, we would still have only one conversion of the VM configuration. That seems like the best approach since Kubernetes works with Yaml formats.  
```

Comment:
@ahadas That too. I guess the main difference is that, while VDSM is the long arm of ovirt-engine in doing its tasks, Kubevirt manages VMs on it's own, it's own way. That for instance also means that a VM start *might* mean that the engine has to push multiple JSONs for different purposes ...

Again not clear enough what I say here. Let my try it again from a different angle: Kubevirt does nothing for ovirt-engine or vdsm and there will not be any oVirt related code in Kubevirt (like this templating thing with the xml where kubevirt would then just fill in some missing data). Kubevirt should have enough logic itself to run VMs, oVirt can just take the role of higher manangement. Finally in case of doubt which way to go, either the oVirt way or the Kubernetes way, Kubevirt will always prefer the Kubernetes way.

Of course that does not mean that Kubevirt can't support oVirt requirements (like providing storage or networks via vdsm, while starting  is done by kubevirt), but we will need generic extensions like [CNI](https://github.com/containernetworking/cni) or [flexVolumes](https://github.com/kubernetes/kubernetes/tree/master/examples/volumes/flexvolume) which can be plugged into KubeVirt. These plugins can the fill in the missing gaps.

To conclude this: I think the milestones @ahadas created here are reasonable, I just find it important to clarify things a little bit that we can see if the proposed ways fit into kubevirt and what needs to be changed there. We don't have to do it in this proposal, but we definitely need then follow up proposals for every milestone, where we can discuss details. But definitely XML plays no role in Kubevirt.